src/main/solution_space/solution_graph.py

```diff
@@ -52,3 +58,97 @@ def add_child(self, child: 'Vertex') -> None:
+        visited = [self._root]
+        vertices_queue = collections.deque(visited)
+        while vertices_queue:
+            vertex = vertices_queue.popleft()
+            for child in vertex.children:
+                if child not in visited and child.vertex_type != VERTEX_TYPE.END.value:
+                    vertices_queue.append(child)
+                    visited.append(child)
+        return visited
+
+    def __next__(self):
+        if self._cursor + 1 >= len(self._traversal):
+            raise StopIteration
+        self._cursor += 1
+        return self._traversal[self._cursor]
+
+
+class SolutionGraph(collections.abc.Iterable):
+    def __init__(self):
+        self._start_vertex = Vertex(vertex_type=solution_space_consts.VERTEX_TYPE.START.value)
+        self._end_vertex = Vertex(vertex_type=solution_space_consts.VERTEX_TYPE.END.value)
+
+    @property
+    def start_vertex(self) -> Vertex:
+        return self._start_vertex
+
+    @property
+    def end_vertex(self) -> Vertex:
+        return self._end_vertex
+
+    def __iter__(self) -> GraphIterator:
+        return GraphIterator(self._start_vertex)
+
+    def get_traversal(self):
+        return self.__iter__().traversal
+
+    def find_or_create_vertex(self, code: Optional[Code], code_info: CodeInfo) -> Vertex:
+        if code is None:
+            log.info('Code should not be None')
+            raise ValueError('Code should not be None')
+        vertices = iter(self)
+        for vertex in vertices:
+            if vertex.code and are_asts_equal(vertex.code.ast, code.ast):
+                log.info(f'Found an existing vertex for code: {str(code)}')
+                vertex.add_code_info(code_info)
+                return vertex
+        log.info(f'Not found any existing vertex for code: {str(code)}, creating a new one')
+        vertex = Vertex(code)
+        vertex.add_code_info(code_info)
+        if code.is_full():
+            log.info(f'Connect full code to the end vertex')
+            self.connect_to_end_vertex(vertex)
+        return vertex
+
+    def connect_to_start_vertex(self, vertex) -> None:
+        self._start_vertex.add_child(vertex)
+
+    def connect_to_end_vertex(self, vertex) -> None:
+        self._end_vertex.add_parent(vertex)
+
+    def add_code_info_chain(self, code_info_chain: List[Tuple[Code, CodeInfo]]) -> None:
+        log.info(f'Start adding code-user chain')
+        if code_info_chain:
+            log.info(f'Connect the first vertex in a chain to the start vertex')
+            first_code_info = code_info_chain[0]
+            first_vertex = self.find_or_create_vertex(first_code_info[0], first_code_info[1])
+            self.connect_to_start_vertex(first_vertex)
+
+            prev_vertex = first_vertex
+            for next_code_user in code_info_chain[1:]:
+                next_vertex = self.find_or_create_vertex(next_code_user[0], next_code_user[1])
+                prev_vertex.add_child(next_vertex)
```

Comment:
Should we use `next_vertex.add_parent(prev_vertex)` if we have created a new vertex? It is never used and we lose some links.
Moreover I think we should call  `add_parent` and `add_child` functions in the `find_or_create_vertex ` function, because it is the  internal actions